### DISCUSSION TITLE
Tag entity

### DISCUSSION THREAD

**Discussion by gjohansson-ST** [home-assistant/architecture#1069] [Entity Models] [answered]
### Context

Currently [tags](https://www.home-assistant.io/integrations/tag/)  is implemented in a manner to use them in automation as scanning is triggering an event in the system.

`mobile_app`, `mqtt` and `esphome` is today capable of  triggering a scanned tag by calling `tag.async_scan_tag(hass, tag_id, device_id)`

Current use in an automation:
```YAML
trigger:
    platform: event
    event_type: tag_scanned
```

### Proposal

Implement a `tag` entity / expand the `tag` integration to also create a `tag` entity for each `tag` registered in it's store.

The `tag` integration will itself create the necessary entities from it's store instead of making an entity component and other integrations (as mentioned above) is creating the entities.

#### Implementation

- Implement a new `tag` entity: `TagEntity` which derives from `EventEntity`.
- Create / update / delete `tag` entities by it's events in the `tag` store.
  - Current UI to create/update/delete `tags` are kept as is.
- The new `tag` entity listens to `EVENT_TAG_SCANNED` for updating it's state (triggered time).
  - As event remains there is nothing breaking with current user implementations.
  - One listener managed from setup, not one per tag.
- Add `tag_id` and `device_id` as additional attributes.

### Alternatives considered

Make `TagEntity` an entity component instead and allow other integrations to implement `tag` entities

#### Why not alternative

Tags (cards, stickers, QR codes etc.) can be consumed by multiple integrations and there needs to be central location for managing tags so therefore the `tag` integration is not suitable as an entity component but instead as a self contained integrations as central location for managing all tags and integrations should only be allowed to scan tags (`tag.async_scan_tag`) and  the `tag` integration will be responsible for triggering events, manage entities and update it's central store.

**Comment by balloob** [comment]
This would mean if we have 100 tags, we add 100 entities and 100 event listeners. This is a burden when we startup the UI and it is something that we will probably render.

What would be the benefit of this change?

**Reply by gjohansson-ST** [reply]
We would have 100 entities for 100 tags but it should be enough with one listener if we manage it all centrally in this integration instead of all entities registering a listener.

I think most of the benefits are a copy of why the `EventEntity` was made so I would say
- User experience. Simpler to use entities than constructing automation from events and also something that you can easily put in dashboards etc. if you want
- Consistency. Why would these not be entities if everything else is going to that direction?
- History / logging
- Use in templates

I think it's worth to link to the discussion of the `EventEntity` as the logic to why it was made is similar here https://github.com/home-assistant/architecture/discussions/929

**Reply by balloob** [reply]
The difference between events in Home Assistant and tags, is that tags have a UI that lists all tags in use, with a last triggered column and an animation when live triggered. That's a whole better user experience than we ever had for events.

**Reply by gjohansson-ST** [reply]
Absolutely and there is nothing here to remove that (will update the proposal to make that clear it's not a replacement).
I think the additional benefits of entities still makes it valid to expand this integration according to the proposal.

**Reply by balloob** [reply]
But what features are we unlocking by making these entities?

**Reply by gjohansson-ST** [reply]
- Use in dashboards
- Use in templates
- History and logging

the others reasons to add this is more about making it more user friendly and improve the user experience as entities are simpler to use than events.

**Reply by janiversen** [reply]
Would it not be possible to generalize this a bit more ?

Having an integration that "convert" events to entities, would help a number of integrations. E.g. the shelly integration works very well until you try to integrate an I4 or HT, because those only generates events. You have to catch the events with an automation and maintain e.g. a sensor.

It would be a lot more user friendly, if you could tell the "event" integration, to convert these events into sensors/switches. And the nice side effect is that it would not burden HA, at least not more than the current automations.

**Reply by emontnemery** [reply]
@janiversen The inputs of a Shelly I4 are exposed as event entities which changes state when buttons are pressed, why would you want to instead make a sensor out of the inputs?

**Reply by janiversen** [reply]
I would like to be able to have the I4 as 4 switches (or sensors, as a poor man's read only switch), so I can use them normally in lovelace.

Right now I have an automation that catches the events and creates a switch entity.

What I tried to write was that a generic event integration, could catch these events and create a switch/sensor entity, as configured by the user.

I have not been able to represent the I4 in lovelace without a "converter" (automation) that catches the event....seems a similar issue like you presented.

**Reply by parautenbach** [reply]
What is your i4's mode set to?

**Reply by janiversen** [reply]
Button, I have tried the different modes, but in neither I got an entity which I could add to lovelace.

Be noted, the shelly integration does receive the events and I catch them. so it works, just not very elegantly.

**Reply by balloob** [reply, marked as answer]
@gjohansson-ST sounds good. Let's do it 👍

**Reply by emontnemery** [reply]
@janiversen This is now way off topic from this architecture discussion, but I don't quite follow. If your Home Assistant core installation is up to date, and you use the built-in shelly integration, the 4 inputs on the I4 will be represented as event entities, where the state is the last time they were pressed.

**Comment by edenhaus** [comment]
Implement with https://github.com/home-assistant/core/pull/115048
